### ISSUE TITLE
Autograd API missing

### ISSUE THREAD

**Issue by joverwey** [LaurentMazare/tch-rs#526]
State of the art GANs often use a [gradient penalty](https://paperswithcode.com/method/r1-regularization) as part of the discriminator loss function. 

[In libtorch you would do it like this:](https://pytorch.org/tutorials/advanced/cpp_autograd.html)

```C++
// Use norm of gradients as penalty
auto grad_output = torch::ones_like(output);
auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output}, /*create_graph=*/true)[0];
auto gradient_penalty = torch::pow((gradient.norm(2, /*dim=*/1) - 1), 2).mean();
```

This API seems to be missing from tch-rs making it impossible to implement recent GANs like [StyleGAN](https://github.com/NVlabs/stylegan3)

From what I can see the native function lives in `torch\csrc\autograd\autograd.h`.

```C++
TORCH_API variable_list grad(
    const variable_list& outputs,
    const variable_list& inputs,
    const variable_list& grad_outputs = {},
    c10::optional<bool> retain_graph = c10::nullopt,
    bool create_graph = false,
    bool allow_unused = false);
```

This should translate to something like;
```Rust
pub fn grad(outputs: &[&Tensor], inputs: &[&Tensor], grad_outputs: &[&Tensor], retain_graph: Option<bool>, create_graph: bool, allow_unused: bool) 
```

How can I add the `torch::autograd::grad` function to tch-rs?

**Comment by joverwey** [comment]
Turns out that you can achieve the same thing with `run_backward`:

```Rust
fn gradient_penalty(input: &Tensor, output: &Tensor) -> Tensor {
    let batch_size = input.size()[0];
    let gradients = Tensor::run_backward(&[output.sum(Kind::Float)], &[input], true, true);
    gradients[0]
        .pow(&2.into())
        .view([batch_size, -1])
        .sum_dim_intlist(&[1], false, Kind::Float)
        .mean(Kind::Float)
}```

**Comment by MilkFather** [comment]
@joverwey Sorry to bother but could you elaborate a bit on this? I am in a pretty similar situation where I have to call `autograd::grad`. In particular I wish to know what's the point of computing the sum of the output first, as shown in your `Tensor::run_backward` solution. Does that mean if I sum the output tensor first then call `Tensor::run_backward`, I will be granted the exact same result as calling `autograd::grad`? PyTorch does not have a `run_backward` function in its docs, and this crate is quite underdocumented. That's a great pain.

**Comment by joverwey** [comment]
If you look at the [libtorch code](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/autograd.cpp), you'll see, that after some initial checks and initialization, autograd::grad just calls run_backward. 

I'm not 100% sure what make_grads do but, to verify, I just called the gradient_penalty above, both in Rust and in the equivalent in PyTorch to verify the results are the same.

**Comment by joverwey** [comment]
Actually, I'm no longer convinced that run_backward and grad does exactly the same thing... In isolation, it seems to compute the same values, but I think there are side-effects that I'm not aware of. I'm getting different performance for my model, depending on whether I train with PyTorch or an equivalent model implemented in `tch-rs`. Maybe my models are not exactly the same, but from my tensorboard results, the gradient penalty loss is different between the two, so it seems likely to be the culprit. I would feel more confindent if `tch-rs` called into the exact same function as PyTorch.
